Cabal/src/Distribution/PackageDescription/Check/Prim.hs

```diff
@@ -0,0 +1,1164 @@
+        | DeclaredUsedFlags (Set.Set FlagName) (Set.Set FlagName)
+        | NonASCIICustomField [String]
+        | RebindableClashPaths
+        | RebindableClashPackageInfo
+        | WErrorUnneeded String
+        | JUnneeded String
+        | FDeferTypeErrorsUnneeded String
+        | DynamicUnneeded String
+        | ProfilingUnneeded String
+        | UpperBoundSetup String
+        | DuplicateModule String [ModuleName]
+        | PotentialDupModule String [ModuleName]
+        | BOMStart FilePath
+        | NotPackageName FilePath String
+        | NoDesc
+        | MultiDesc [String]
+        | UnknownFile String (SymbolicPath PackageDir LicenseFile)
+        | MissingSetupFile
+        | MissingConfigureScript
+        | UnknownDirectory String FilePath
+        | MissingSourceControl
+        | MissingExpectedDocFiles Bool [FilePath]
+        | WrongFieldForExpectedDocFiles Bool String [FilePath]
+    deriving (Eq, Ord, Show)
+    -- TODO Some checks have a constructor in list form
+    --      (e.g. `SomeWarn [n]`), CheckM m () correctly catches warnings in
+    --      different stanzas in different checks (so it is not one soup).
+    --
+    --      Ideally [SomeWar [a], SomeWar [b]] would be translated into
+    --      SomeWar [a,b] in the few cases where it is appropriate for UX
+    --      and left separated otherwise.
+    --      To achieve this the Writer part of CheckM could be modified
+    --      to be a ad hoc monoid.
+
+-- Convenience.
+extractCheckExplantion :: PackageCheck -> CheckExplanation
+extractCheckExplantion (PackageBuildImpossible e) = e
+extractCheckExplantion (PackageBuildWarning e) = e
+extractCheckExplantion (PackageDistSuspicious e) = e
+extractCheckExplantion (PackageDistSuspiciousWarn e) = e
+extractCheckExplantion (PackageDistInexcusable e) = e
+
+-- | Which stanza does `CheckExplanation` refer to?
+--
+data CEType = CETLibrary | CETExecutable | CETTest | CETBenchmark
+    deriving (Eq, Ord, Show)
+
+-- | Pretty printing `CEType`.
+--
+ppCE :: CEType -> String
+ppCE CETLibrary = "library"
+ppCE CETExecutable = "executable"
+ppCE CETTest = "test suite"
+ppCE CETBenchmark = "benchmark"
+
+-- | Which field does `CheckExplanation` refer to?
+--
+data CEField = CEFCategory | CEFMaintainer | CEFSynopsis
+             | CEFDescription | CEFSynOrDesc
+    deriving (Eq, Ord, Show)
+
+-- | Pretty printing `CEField`.
+--
+ppCEField :: CEField -> String
+ppCEField CEFCategory = "category"
+ppCEField CEFMaintainer = "maintainer"
+ppCEField CEFSynopsis = "synopsis"
+ppCEField CEFDescription = "description"
+ppCEField CEFSynOrDesc = "synopsis' or 'description"
+
+-- | Which language are we referring to in our warning message?
+--
+data WarnLang = LangC | LangCPlusPlus
+    deriving (Eq, Ord, Show)
+
+-- | Pretty printing `CheckExplanation`.
+--
+ppWarnLang :: WarnLang -> String
+ppWarnLang LangC = "C"
+ppWarnLang LangCPlusPlus = "C++"
+
+-- | Pretty printing `CheckExplanation`.
```

Comment:
I think, from here on you could go into a separate module. Something like `Distribution.PackageDescription.Check.Explanations`

Reply:
Legacy `Show` instance for `CheckExplanation` would create an import loop. Once we deprecate it and reclaim it to a sane instance (show ⇄ read), we can have this. I must admit while working on `check` I liked the ergonomics of having pretty printer near the type.

Reply:
Makes sense. Just this part looks sufficiently monolithic to me. But we don't need loops for sure!